contracts/integrations/contracts/dydx/ZeroExV3ExchangeWrapper.sol

```diff
@@ -0,0 +1,329 @@
+        if (protocolFeeCollector != address(0)) {
+            WETH_TOKEN.ensureAllowance(
+                protocolFeeCollector,
+                MathHelpers.maxUint256()
+            );
+        }
+
+        // make sure that the exchange can take the tokens from this contract
+        takerToken.ensureAllowance(
+            ZERO_EX_TOKEN_PROXY,
+            requestedFillAmount
+        );
+
+        // do the exchange
+        LibFillResults.FillResults memory fill = v3Exchange.fillOrKillOrder(order, requestedFillAmount, signature);
+
+        // validate fill amount
+        assert(fill.takerAssetFilledAmount == requestedFillAmount);
+
+        // validate protocol fee
+        validateProtocolFee(fill.protocolFeePaid);
+
+        // set allowance
+        makerToken.ensureAllowance(
+            receiver,
+            fill.makerAssetFilledAmount
+        );
+
+        return fill.makerAssetFilledAmount;
+    }
+
+    function getExchangeCost(
+        address makerToken,
+        address takerToken,
+        uint256 desiredMakerToken,
+        bytes calldata orderData
+    )
+        external
+        view
+        returns (uint256)
+    {
+        (LibOrder.Order memory order, ) = parseOrder(orderData, makerToken, takerToken);
+
+        return MathHelpers.getPartialAmountRoundedUp(
+            order.takerAssetAmount,
+            order.makerAssetAmount,
+            desiredMakerToken
+        );
+    }
+
+    function getMaxMakerAmount(
+        address makerToken,
+        address takerToken,
+        bytes calldata orderData
+    )
+        external
+        view
+        returns (uint256)
+    {
+        (LibOrder.Order memory order, ) = parseOrder(orderData, makerToken, takerToken);
+        IExchange v3Exchange = IExchange(ZERO_EX_EXCHANGE);
+        LibOrder.OrderInfo memory orderInfo = v3Exchange.getOrderInfo(order);
+
+        if (orderInfo.orderStatus != uint8(LibOrder.OrderStatus.FILLABLE)) {
+            return 0;
+        }
+
+        uint256 remainingTakerAssetAmount =
+            order.takerAssetAmount.sub(orderInfo.orderTakerAssetFilledAmount);
+
+        return MathHelpers.getPartialAmount(
+            remainingTakerAssetAmount,
+            order.takerAssetAmount,
+            order.makerAssetAmount
+        );
+    }
+
+    // ============ Private Functions ============
+
+    function transferTakerFee(
+        LibOrder.Order memory order,
+        address tradeOriginator,
+        uint256 requestedFillAmount,
+        address takerFeeToken
+    )
+        private
+    {
+        uint256 takerFee = MathHelpers.getPartialAmount(
+            requestedFillAmount,
+            order.takerAssetAmount,
+            order.takerFee
+        );
+
+        if (takerFee == 0) {
+            return;
+        }
+
+        require(
+            TRUSTED_MSG_SENDER[msg.sender],
+            "ZeroExV3ExchangeWrapper#transferTakerFee: Only trusted senders can dictate the fee payer"
+        );
+
+        takerFeeToken.transferFrom(
```

Comment:
So it is assumed that this contract has a WETH balance, correct? Given that this transfer happens after the fill is executed.

Reply:
yes, that's the assumption